File: java/sort/linked_list/InsertionSort.java
package sort.linked_list;

public class InsertionSort {
    public void sort(ListNode head) {
        ListNode current = head.next;
        while (null != current.next) {
            ListNode prev = head;
            while (prev != current && prev.next.value <= current.next.value) {
                prev = prev.next;
            }

            if (prev == current) {
                current = current.next;
            } else {
                ListNode next = current.next;
                current.next = next.next;
                next.next = prev.next;
                prev.next = next;
            }
        }
    }

    public class ListNode {
        public int value = 0;
        public ListNode next = null;

        public ListNode(int value) {
            this.value = value;
        }
    }
}

File: py/sort/linked_list/InsertionSort.py
#!/usr/bin/env python
# -*- coding:gbk -*-

class ListNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class InsertionSort:
    def sort(self, head: ListNode) -> None:
        current = head.next
        while current.next:
            prev = head
            while prev != current and prev.next.value <= current.next.value:
                prev = prev.next

            if prev == current:
                current = current.next
            else:
                next = current.next
                current.next, next.next, prev.next = next.next, prev.next, next
